Request Scheduler from 3d-tiles

lilleyse opened the pull request:
For #3241 

Bringing in the Request Scheduler into master. 

I also brought in `TileBoundingVolume` and its sublcasses `TileBoundingRegion`, `TileBoundingSphere`, and `TileOrientedBoundingBox` into master because they are being used to calculate tile distance in `GlobeSurfaceTile` which gets passed to the request scheduler.

All the miscellaneous request functions now pass through the `RequestScheduler`.

**TODO**

* [x] Update CHANGES.md

pjcozzi commented on CHANGES.md:
```diff
@@ -8,6 +8,7 @@ Change Log
 * Deprecated
     * Deprecated `GroundPrimitive.geometryInstance`. It will be removed in 1.20. Use `GroundPrimitive.geometryInstances` instead.
     * Deprecated `TileMapServiceImageryProvider`. It will be removed in 1.20. Use `createTileMapServiceImageryProvider` instead.
+    * Deprecated `throttleRequests` flag in `TerrainProvider`. It will be removed in 1.19. It is replaced by an optional `Request` object that stores information used to prioritize requests.
```
I just noticed that `Request` is still private.  Perhaps let's not deprecate this until we make the `RequestScheduler` public.  If you agree, update the 3D Tiles roadmap.

pjcozzi replied:
Or maybe `Request` should be public.  Can we reasonably do that without making the entire request scheduler public since I don't think we want to commit to the full public API today.

pjcozzi replied:
@lilleyse what is the plan here?  See #3448.

Also, move this to the 1.25 section of CHANGES.md.

lilleyse commented:
@pjcozzi This is up-to-date with master and the new spec changes.

pjcozzi commented:
@lilleyse can you evaluate #4166 before we merge this?

pjcozzi commented:
Wow, congratulations on 100% code coverage in the new files.

The code all looks good.  This just needs #4166 and #3448.

mramato commented:
I may have brought this up before, but are we sure this is going to be a net win for us?  It seems like we are trying to be smarter than the browser at its own game.  I can understand prioritizing requests based on scene geometry (requests important tiles/imagery/terrain before further away stuff), but doing our own throttling seems like a really bad idea, especially with HTTP/2 adoption starting to happen.  (where Cesium will end up a lot slower than it should be out of the box because no one is going to know to adjust these settings).

I also imagine this system breaks down if it is part of a larger app where requests are being made outside of the RequestScheduler (since the browser will end up throttling anyway).

I'm sure there's been a lot of thought and offline discussion put into this, and I don't mean to step on anyone's toes; I just want to understand the big picture here and be sure that this is a big enough win for us that it's worth the extra complexity.

pjcozzi commented:
We definitely need system-wide (cross terrain, imagery, 3D tilesets, etc.) priority for requests.  We can't have, for example, terrain requests in the distance starving 3D building requests in the foreground.

As for throttling based on subdomains, I think it is fine to take this change with the same behavior as our current throttling.  If I understand, this change correctly it just makes the throttling system-wide.  Once whatever issues are worked out, I would be fine merging this re-evaluating as HTTP/2 spreads.  We'll always want some kind of limit so, for example, we don't make 10K requests only to realize the next frame that we don't actually need any of them and instead want just a new 10 requests.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,440 @@
+     * @type {Number}
+     * @default 6
+     */
+    RequestScheduler.maximumRequestsPerServer = 6;
```
How did we choose this number, because that's what Chrome does?  I believe other browsers have different values.  If this number is really important, should we have some browser-specific test code in here to determine a better default?

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,440 @@
+     * @type {Number}
+     * @default 10
+     */
+    RequestScheduler.maximumRequests = 10;
```
Same question as above.

mramato commented:
> We definitely need system-wide (cross terrain, imagery, 3D tilesets, etc.) priority for requests. We can't have, for example, terrain requests in the distance starving 3D building requests in the foreground.

I agree, but I would really like to see some hard numbers in a variety of benchmarks across all of the major browsers comparing this branch and master to prove we aren't making things worse.  When it comes to performance, we've always been about hard numbers in the past, no reason to stop now.

pjcozzi commented:
Yes, I'd like to see some numbers too.  This won't be better than master, but it not be non-trivially worse.

mramato commented:
@lilleyse should this just be closed?  Or do you plan on reworking it before opening the `3d-tiles` PR?

lilleyse commented:
> @lilleyse should this just be closed? Or do you plan on reworking it before opening the 3d-tiles PR?

I plan on reworking it with @austinEng's help, but will open the `3d-tiles` PR before hand and link back to here. It can stay open for now.

lilleyse commented:
This is almost ready but I still need to add tests and do benchmarks.

I worked from a bunch of the ideas here (https://github.com/AnalyticalGraphicsInc/cesium/pull/5317). Per-frame budgets are now gone so the problems when using multiple widgets https://github.com/AnalyticalGraphicsInc/cesium/issues/5226 should be fixed.

A request now contains a state enum:

```
    var RequestState = {
        UNISSUED : 0,   // Initial unissued state.
        ISSUED : 1,     // Issued but not yet active. Will become active when open slots are available.
        ACTIVE : 2,     // Actual http request has been sent.
        DONE : 3,       // Request completed successfully.
        CANCELLED : 4,  // Request was cancelled, either explicitly or automatically because of low priority.
        FAILED : 5      // Request failed.
    };
```
When terrain/imagery/3d-tiles makes a request, a few things can happen:
* Returns `undefined` instead of a promise if too many requests are active. Same behavior as before.
* Returns a promise
    * Promise is resolved when the `XMLHttpRequest` is complete
    * Promise is rejected if the request was cancelled. This can happen if the request is explicitly cancelled (only in `3d-tiles` traversal right now) or the request was kicked off the `requestHeap` before the xhr was sent.
        * `TileTerrain` and `ImageryLayer` check for cancelled requests and reset the state of the tile/imagery so that it can request again.

Other changes:
* `RequestScheduler.request` and `RequestScheduler.schedule` are now the same thing and just called `RequestScheduler.request`. A request may be marked as `throttled`, meaning it will be put through the `requestHeap` and wait an open slot in `activeRequests`. It may be cancelled if other requests come in that are higher priority. Requests that are not throttled, which is almost everything except terrain/imagery, go through immediately.
* Storing the xhr object in the `Request` object became really complicated with the old syntax. Now requests do not go through `RequestScheduler` directly. Instead `loadWithXhr`, `loadImage`, `loadJsonp`, and subordinates call `RequestScheduler` internally. Each of these functions takes a `Request` object as its last parameter. This means there are a lot less changes throughout the codebase.
* More breaking changes - `requestTileGeometry` no longer has a `throttle` argument, this is replaced by a `Request`.
* I'm not 100% solid on my changes in `GoogleEarthEnterpriseMetadata`. I may need to discuss with @tfili about this.
* I removed throttling by subdomain, but may add this back.

If anyone wants to give this a quick sanity-check that would be nice, but otherwise I still have more to do to get this ready and will update then.

pjcozzi commented:
> This is almost ready but I still need to add tests and do benchmarks.

As part of this, can you please spot check a few representative views for terrain/imagery in this branch vs master to verify the number and order of requests are about the same?

pjcozzi commented:
> DONE : 3,       // Request completed successfully.

`RECEIVED` is probably better semantics and more consistent with other parts of Cesium.

pjcozzi commented:
> RequestScheduler.request and RequestScheduler.schedule are now the same thing and just called RequestScheduler.request

Why not remove `schedule`?

pjcozzi commented:
> I removed throttling by subdomain, but may add this back.

Seems like we would want this.

pjcozzi commented:
Added a **TODO** section to the first comment in this PR.

pjcozzi commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -532,17 +531,18 @@ define([
 
-        function tileLoader(tileUrl) {
-            return loadArrayBuffer(tileUrl, getRequestHeader(extensionList));
+        // TODO - remove later, this handles the deprecated throttleRequests parameter
```
Remove the TODO and submit an issue to deprecate this.

Same comment throughout.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,114 @@
+         * Whether to throttle and prioritize the request. If false, the request will be sent immediately. If true, the
+         * request will be throttled and sent based on priority.
+         *
+         * @private
```
The design of this class is atypical.  If the constructor function takes a parameter, that parameter can generally be returned by a property.  Are we sure these should be `private`?  Perhaps they should be readonly getters?

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,114 @@
+        this.screenSpaceError = defaultValue(options.screenSpaceError, 0.0);
+
+        /**
+         * The URL to request. Set by the following load functions:
```
The built doc will look strange since the `@see` tags are in a separate section.  Reword or include them in the sentence with `@link` tags.

pjcozzi replied:
Also, this is somewhat strange behavior: should the constructor function tell users not to construct this object directly and instead call one of these function?

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,114 @@
+        this.url = undefined;
+
+        /**
+         * The actual function that makes the request. The function takes no arguments and returns a promise for the requested data.
```
Same comment.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,256 @@
+     * @type {Number}
+     * @default 50
+     */
+    RequestScheduler.maximumRequests = 50;
```
I imagine this is part of your ongoing tuning, but are we sure about this default?  And how do users configure this since the class is private?

lilleyse replied:
For now this option doesn't actually have much effect since terrain/imagery providers are throttled to 6 per server. This is more of a safety net if we decide to eventually remove the `throttleByServer` option for http2 providers so as not to overload the browser.

For now this class and these options should just stay private.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,256 @@
+     * @type {Number}
+     * @default 20
+     */
+    RequestScheduler.priorityHeapSize = 20;
```
My original idea was to queue up all the requests in a frame, "sort" by most important, and then issue as many as possible to make best use of the bandwidth and our ability to process them once they are downloaded.

I actually think this can create bubbles in the requests, e.g., it would be good to make some best guest high priority requests even during the frame as requests come back in.

Is that how it works now?  I guess I will see...

lilleyse replied:
Yeah that's what happens here. One thing I'm curious about is what the browser does under the hood. Does it also queue requests until the end of the frame?

pjcozzi replied:
I dunno, I suspect not, but let me know what you find out.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,256 @@
+     * @type {Boolean}
+     * @default true
+     */
+    RequestScheduler.throttle = false;
```
Is this just for debugging?  If so, prefix prefix with `debug` like the graphics code.

pjcozzi commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,256 @@
+
+    function sortRequests(a, b) {
+        // Sort requests by higher screen space error and then closer distance
+        if (a.screenSpaceError !== b.screenSpaceError) {
```
Should the constructor function have any validation for making SSE/distance either/or?

Are we sure that this is even the right metric?  Maybe distance and how center the tile is in the viewport is more important than SSE, which would include distant tiles with large error.  Or maybe this is biased to distance/centeredness, but let's through `n` lower priority requests every so often to prevent stalling them but keep the most important detail coming in.

pjcozzi replied:
Also, not scoped for this PR, but I wanted to experiment with effectively increasing the SSE when the camera is moving quickly (e.g., for camera flights) so tiles are not requested and then out of view when they are returned.  Perhaps the request scheduler will be the right place to add this logic since it would work system-wide and still allow the tileset/terrain/imagery to render with high SSE if it is already loaded.

lilleyse replied:
I removed the SSE sorting for now because it is not used by terrain/imagery. I'll evaluate this again when merging the request scheduler back into 3d-tiles.

pjcozzi commented on Source/Core/isBlobUri.js:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @private
+     */
+    function isBlobUri(uri) {
```
Why not
```javascript
return defined(uri) && blobUriRegex.test(uri);
```

mramato replied:
Also, is it normal for `uri` to be undefined?  Why wouldn't that be a `DeveloperError`?

pjcozzi commented:
What's the plan with this w.r.t to HTTP2 and #5316?

pjcozzi commented:
>> RequestScheduler.request and RequestScheduler.schedule are now the same thing and just called RequestScheduler.request
>
> Why not remove schedule?

ha, never mind, I get it now.

pjcozzi commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,114 @@
+     * @param {Number} [options.distance=0.0] The distance from the camera, used to prioritize requests.
+     * @param {Number} [options.screenSpaceError=0.0] The screen space error, used to prioritize requests.
+     */
+    function Request(options) {
```
Does this and `RequestType` need to be public or can this whole request scheduling happen behind the scenes for now?

lilleyse replied:
I'm starting to think it can all be private.

mramato replied:
`requestTileGeometry` calls are definitely not private.  They are part of the publicly documented interface for `TerrainProvider`, same with `requestImage` and `ImageryProvider` so `Request`, and `RequestType` definitely need to be public.

mramato commented:
I think the answer is no, but is this a breaking change for custom terrain and imagery providers?  At the very least, it's not breaking but custom providers should opt-in to throttling.

lilleyse commented:
I'm starting to see some performance benefits here (rough averages):
```
Static Horizon View
    master
        9.89 seconds
        262 requests
    branch
        9.79 seconds
        255 requests
```

```
Zoom From Space:
    master
        6.41 seconds
        216 requests
    branch
        5.93 seconds
        190 requests
```

Code should be updated pretty soon.

mramato commented:
@lilleyse what's the status here?  Is this ready for final review? (CI fixes aside)

lilleyse commented:
@mramato I'm still writing and fixing tests, but if you want an early start on the review I don't expect much to change.

lilleyse commented on CHANGES.md:
```diff
@@ -2,9 +2,11 @@ Change Log
    * `GoogleEarthImageryProvider` has been deprecated and will be removed in Cesium 1.37, use `GoogleEarthEnterpriseMapsProvider` instead. 
-* 
+   * The `throttleRequest` parameter for `TerrainProvider.requestTileGeometry`, `CesiumTerrainProvider.requestTileGeometry`, `VRTheWorldTerrainProvider.requestTileGeometry`, and `EllipsoidTerrainProvider.requestTileGeometry` is deprecated and will be replaced with an optional `Request` object. The `throttleRequests` parameter will be removed in 1.36, instead to throttle requests set the request's ...
+   * The ability to provide a Promise for the `options.url` parameter of `loadWithXhr` is deprecated. The same applies for the `url` parameter for `loadArrayBuffer`, `loadBlob`, `loadImageViaBlob`, `loadText`, `loadJson`, `loadXML`, `loadImage`, `loadCRN`, `loadKTX`, and `loadCubeMap`. This will be removed in 1.36, instead `url` must be a string.
```
I checked that passing a promise instead of a string is not used by any of the tests. There could be old code or demos that rely on this though. 

Also the number `1.36` is just placeholder, what would be good times to remove?

lilleyse replied:
Also a request object is now the last parameter for every `load` related function. Is this a change worth documenting here?

pjcozzi replied:
> Also the number 1.36 is just placeholder, what would be good times to remove?

Probably two releases so 1.37.

> Also a request object is now the last parameter for every load related function. Is this a change worth documenting here?

Here or below in CHANGES.md.

mramato replied:
Sorry if this has already been asked and answered, but why remove this functionality?  Just because we think no one is using it?  Ultimately I'm okay with it.

lilleyse replied:
Actually I didn't really explain this. In order to throttle by server the `RequestScheduler` needs to derive the server from the string url.

This used to be fine, just do a `when(url)` call first. But to signal that a request was throttled the load function (via `RequestScheduler`) need to be able to return `undefined`.

mramato replied:
Thanks, makes sense.

lilleyse commented on Source/Core/CesiumTerrainProvider.js:
```diff
@@ -489,17 +493,16 @@ define([
      *            returns true.
      */
-    CesiumTerrainProvider.prototype.requestTileGeometry = function(x, y, level, throttleRequests) {
+    CesiumTerrainProvider.prototype.requestTileGeometry = function(x, y, level, request) {
```
Right now every terrain/imagery provider sets `throttleByServer` to true which limits to 6 active requests per server. Once more of these servers switch over to http2 this flag will hold them back. With `CesiumTerrainProvider` and some others we'll know when the make the switch, but this doesn't help people using custom providers. Should an option like `throttleByServer` be exposed on the provider level for users to tweak?

This same idea applies to 3D tilesets too. A `throttleByServer` option could be sent in when constructing the tileset.

pjcozzi replied:
I think something less coarse-grained or at least named differently needs to be exposed.  Have you looked at other engines using HTTP2 or doc on best practices with HTTP2?

mramato replied:
I think I've been pretty clear that I'm squarely in the "don't throttle" crowd (I'm all for prioritization and cancellation, just not throttling) so take this with that bias up front.  

Just because a server supports HTTP/2, doesn't mean the client does.  So it's impossible for the developer to make this decision during development.  As far as I know, it's also impossible to detect at runtime whether you're using HTTP/2, so we can't do some magic on our end either.  That means the best option is to not throttle at all and let the browser handle it.

I tried to hunt around to see what other webmaps are doing, and as far as I can tell none of them throttle (I checked OL/Leaflet/Mapbox).  OpenLayers has an open issue to cancel unneeded requests, but that's it.

But since I'm not going to be able to convince you guys to not do it, I think the second best option is having a single top-level setting on `RequestScheduler` to disable it.  I know the initial implementation had this, I'm not sure about the one in this branch (haven't gotten to that part of the code yet).

lilleyse replied:
If HTTP/2 becomes widely adopted, I agree that throttling by server is a waste. However as of now throttling does make a big difference to performance:

I just got these numbers for an Everest static view:
```
4G with cache disabled:
  Throttled:
     9.70 seconds
     256 requests
  Not throttled:
     15.17 seconds
     412 requests

Normal usage
  Throttled
    1.45 seconds
    526 requests
  Not throttled
    1.60 seconds
    526 requests
```

When would the client not support HTTP/2?

mramato replied:
> When would the client not support HTTP/2?

Up until recently, not all browsers supported it, and even then there could be lots of atypical browsers (embedded, webview, etc..) that don't support it.  In the long run, everyone will.

> I just got these numbers for an Everest static view:

I know I'm missing something, but why are there less requests with the 4G version?  Shouldn't the total number of requests be the same? (if we're just talking about throttling and not cancellation).

pjcozzi replied:
I suspect that none of us individually knows enough about everything - tile selection and HTTP2 - to make a good call here.  Let's get together for a few minutes today in person.  @lilleyse please organize.

mramato commented on Source/Core/Heap.js:
```diff
@@ -0,0 +1,200 @@
+        Check.defined('comparator', comparator);
+        //>>includeEnd('debug');
+
+        this._comparator = comparator;
```
In general, you should always expose any constructor parameters as properties (in this case a public getter would be fine).  People should always be able to get back out of an object what they put into it.

mramato commented on Source/Core/Heap.js:
```diff
@@ -0,0 +1,200 @@
+         * @type {Array}
+         * @readonly
+         */
+        internalArray : {
```
Kind of odd to expose this, but since it's a private class, I guess I'm okay with it.

mramato commented on Source/Core/Heap.js:
```diff
@@ -0,0 +1,200 @@
+        var array = this._array;
+        var candidate = -1;
+
+        while (true) {
```
Can we change this to a `do...while` loop instead?  I'm not a big fan of `while(true)`.

lilleyse replied:
I've tried other ways but this just seems the clearest to me.

mramato replied:
OK, but why wouldn't a `do while` loop make more sense here?

mramato commented on Source/Core/Heap.js:
```diff
@@ -0,0 +1,200 @@
+     * Insert an element into the heap. If the length would grow greater than maximumLength
+     * of the heap, extra elements are removed.
+     *
+     * @param {*} element The element to insert
```
`*` -> `Object` (and below).

mramato replied:
Actually, I'm wrong here.  `*` is correct.  I wrote up #5411 to clear this up throughout the codebase.

mramato commented on Source/Core/Heap.js:
```diff
@@ -0,0 +1,200 @@
+
+        var removedElement;
+
+        if (defined(this._maximumLength) && (this._length > this._maximumLength)) {
```
both `this._length` and `this._maximumLength` are accessed multiple times throughout this function, store them in locals.

mramato commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,162 @@
+         *
+         * @type {Function}
+         *
+         * @private
```
Why is this private?  It was passed into the constructor.

mramato replied:
All of these public properties should be marked readonly as well, right?

lilleyse replied:
This and other comments are my overall dilemma with the `Request` class. It needs to be exposed because it is now added to `requestTileGeometry` the `load` functions, but it is not meant to be constructed directly. The one exception might be if someone is calling `requestTileGeometry` but wants to set `throttle : true`. I'm trying to hide the other stuff as much as possible. This is why the documentation also seems conflicting.

mramato commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,162 @@
+         *
+         * @default 0.0
+         */
+        this.distance = defaultValue(options.distance, 0.0);
```
This is a leaky abstraction.  It's odd to have `distance` and `screenSpaceError` on the Request object.  I understand that there needs to be a metric to prioritize request, but why not just have this be a function that returns a unitless priority value and then standard implementations for world-based objects?

I can also imagine other odd situations where you are loading 100 instances of the same model.  Who's position do you use for the distance calculation in the queue?  My turning this into a callback, each RequestScheduler user can optimize these calls for their needs.

mramato commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,162 @@
+         *
+         * @private
+         */
+        this.server = undefined;
```
What exactly does this mean?  Because a `server` can mean a lot of different things.  Is it the full url?  Just the domain?  Does it include the port and any other information?

mramato commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,162 @@
+         *
+         * @private
+         */
+        this.xhr = undefined;
```
Why are the private properties here not prefixed with an underscore?  Also, why not just issue the abort in the cancel call below?

mramato replied:
I'm also a little concerned with leaky abstraction again here.  Since we want more robust Node support long term, letting the xhr instance "escape" the `loadWithXhr` call may cause headaches down the road.  Not a show stopper right now, but something we may want to consider looking into sooner rather than later post-merge.

lilleyse replied:
> why not just issue the abort in the cancel call below?

I try to keep most of the request handling in `RequestScheduler.update`.

> I'm also a little concerned with leaky abstraction again here. Since we want more robust Node support long term, letting the xhr instance "escape" the loadWithXhr call may cause headaches down the road. Not a show stopper right now, but something we may want to consider looking into sooner rather than later post-merge.

Hm.. there could be a `cancelRequest` callback instead of storing the `xhr` here.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+     *
+     * @exports RequestScheduler
+     *
+     * @private
```
So this is private now?  Don't we need a way to disable throttling and configure other parameters?

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+    /**
+     * The maximum number of simultaneous active requests per server. Un-throttled requests do not observe this limit.
+     * @type {Number}
+     * @default 50
```
`50` -> `6`

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+     * @type {Boolean}
+     * @default true
+     */
+    RequestScheduler.debugThrottle = true;
```
I would just call this `throttleRequests` instead of `debugThrottle`.

mramato commented on Source/Core/Request.js:
```diff
@@ -0,0 +1,162 @@
+         *
+         * @default 0.0
+         */
+        this.screenSpaceError = defaultValue(options.screenSpaceError, 0.0);
```
As far as I can tell, this isn't used anywhere.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+        }
+    });
+
+    function sortRequests(a, b) {
```
Normally, we would put everything from here until 107 at the top of the file.  Kind of weird having it in the middle like this (especially since statistics is used above.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+        numberOfActiveRequestsEver : 0
+    };
+
+    var requestHeap = new Heap(sortRequests);
```
Why not just have `Heap` take an `options` object?  Would make this code cleaner.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+        var uri = new Uri(url).resolve(pageUri);
+        uri.normalize();
+        var serverName = uri.authority;
+        if (!/:/.test(serverName)) {
```
Took me a minute to figure out what was going on here.  A comment would go a long way.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+    };
+
+    /**
+     * For testing only.
```
I would put all "testing only" stuff at the very bottom of the file.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+        }
+        request.state = RequestState.CANCELLED;
+        ++statistics.numberOfCancelledRequests;
+        request.deferred.reject('Cancelled');
```
It's odd to reject with a string like this.

mramato replied:
I'll also add that Promise cancellation is a thing that exists and is in Bluebird.  When we finally get around to switching Cesium over, we may be able to use that here instead of rejecting.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+     * Issuers of a request should update properties of requests. At the end of the frame,
+     * RequestScheduler.update is called to start, cancel, or defer requests.
+     */
+    RequestScheduler.update = function() {
```
Since this a static object, but called by the scene.  Is there any concern with calling this once for each frame per-scene instead of once overall?  Would it be better to have `RequestScheduler` use it's own `requestAnimationFrame` callback? This way it gets called once were animation frame for all scenes.

lilleyse replied:
Multiple viewers/scenes won't be a problem. The update call will just be ineffective for any additional scenes. I tested out https://github.com/AnalyticalGraphicsInc/cesium/issues/5226 on this branch and didn't have any problems.

mramato commented on Source/Core/RequestScheduler.js:
```diff
@@ -0,0 +1,383 @@
+        ++statistics.numberOfAttemptedRequests;
+
+        var isDataOrBlobUri = isDataUri(request.url) || isBlobUri(request.url);
+        if (!isDataOrBlobUri && !defined(request.server)) {
```
If it's a data uri or blob, why put it in the queue at all?  You could just move this check up and call `requestFunction` immediately instead of waiting to the end of the frame.  Is there a hidden benefit to threading it through the system?

lilleyse replied:
The request was going through immediately, but I think the code was just confusing. I've reorganized it. As a general rule, only throttled requests wait for the end of the frame.

mramato commented on Source/Core/RequestState.js:
```diff
@@ -0,0 +1,21 @@
+    'use strict';
+
+    /**
+     * @private
```
I mentioned this already, but I'm not convinced we can get away with making these private.  Can someone implement their own TerrainProvider or ImageryProvider without using these objects?

mramato commented on Source/Core/loadArrayBuffer.js:
```diff
@@ -13,10 +13,10 @@ define([
      * @param {Object} [headers] HTTP headers to send with the requests.
-     * @returns {Promise.<ArrayBuffer>} a promise that will resolve to the requested data when loaded.
-     *
+     * @param {Request} [request] The request object.
```
We have this identical `@param {Request} [request] The request object.` documentation everywhere but the Request object itself says `Stores information for making a request. This should not be constructed directly. Instead, call the appropriate load function, like `loadWithXhr`, `loadImage`, etc.`  One of them is out of date.  Documentation as a whole could be improved.

mramato commented:
That's all I have for this go around @lilleyse.  This is definitely a lot cleaner than the initial version and I look forward to seeing your benchmarks.

I did not review the specs, but I'll do that along with any other changes once you give the go ahead.

Thanks.

mramato commented on Specs/spec-main.js:
```diff
@@ -136,6 +136,8 @@
         window.it = function(description, f, timeout, categories) {
             originalIt(description, function(done) {
+                // Clear the RequestScheduler before for each test in case requests are still active from previous tests
+                Cesium.RequestScheduler.clearForSpecs();
```
Is this still needed here?  I would prefer to remove it unless there's a really good reason for it.

lilleyse replied:
I'll see what I can do, because I don't like this either. Maybe this just needs to be called for specs that are doing more fine-grained request checking.

mramato replied:
Right, it should only matter for tests that are testing the `RequestScheduler` itself.

lilleyse replied:
There could be cases where unrelated tests do not resolve their promises correctly so the `activeRequests` list can get clogged. But since this only affects throttled requests the `clearForSpecs` only need to go in terrain/imagery provider tests.

lilleyse commented:
Results on release build of **master**. All tests disable the browser's cache. The HTTP2 server I'm using hosts both terrain and imagery at the same server, which is not ideal for throttling.

A few takeaways here:
* Throttling works better for slower connections. When using 4G speeds, throttling by 6 per server is a big win especially in the Zoom case. However when using normal speeds, throttling can hurt performance slightly but usually results in less requests overall.
* The http2 results are consistent with above, but visually there is a difference. Without throttling, all the requests come in at the same time so there are these large sweeps where the terrain refines. With throttling, the loading is more incremental and looks better to me.

```
Results

Master | HTTP2 server | no-throttle | Static
    4.86 seconds | 449 requests
    5.07 seconds | 466 requests
    5.04 seconds | 445 requests
    5.26 seconds | 454 requests

Master | HTTP2 server | throttle 6 per-domain | Static
    5.41 seconds | 346 requests
    5.10 seconds | 351 requests
    5.46 seconds | 349 requests
    5.84 seconds | 348 requests

Master | HTTP1 server | no-throttle | Static
    2.03 seconds | 524 requests
    2.19 seconds | 524 requests
    2.38 seconds | 524 requests
    2.70 seconds | 524 requests
    2.98 seconds | 524 requests
    2.98 seconds | 524 requests
    2.64 seconds | 524 requests
    2.73 seconds | 524 requests

Master | HTTP1 server | throttle 6 per-domain | Static
    2.22 seconds | 482 requests
    3.58 seconds | 509 requests
    3.20 seconds | 506 requests
    3.15 seconds | 490 requests
    2.95 seconds | 508 requests
    2.88 seconds | 481 requests
    2.27 seconds | 471 requests
    3.00 seconds | 492 requests

Master | HTTP1 server | no-throttle | Zoom
    3.50 seconds | 825 requests
    3.78 seconds | 793 requests
    3.79 seconds | 776 requests

Master | HTTP1 server | throttle 6 per-domain | Zoom
    3.59 seconds | 802 requests
    3.83 seconds | 750 requests
    3.92 seconds | 750 requests
    3.63 seconds | 821 requests

=======================================

Master | HTTP1 server | no-throttle | Static | 4G
    21.24 seconds | 556 requests
    20.81 seconds | 540 requests

Master | HTTP1 server | throttle 6 per-domain | Static | 4G
    13.41 seconds | 349 requests
    13.69 seconds | 353 requests
    13.45 seconds | 348 requests

Master | HTTP1 server | no-throttle | Zoom | 4G
    33.02 seconds | 1022 requests
    31.89 seconds | 998 requests

Master | HTTP1 server | throttle 6 per-domain | Zoom | 4G
    9.68 seconds | 296 requests
    7.60 seconds | 270 requests

Master | HTTP2 server | no-throttle | Static | 4G
    16.53 seconds | 447 requests
    17.87 seconds | 476 requests

Master | HTTP2 server | throttle 6 per-domain | Static | 4G
    12.52 seconds | 331 requests
    13.39 seconds | 349 requests
```

Will have more stats for this branch to come.

lilleyse commented:
> >I removed throttling by subdomain, but may add this back.
>
>What does this mean? We no are foo.server.com and bar.server.com considered the same thing? If so, I would suggest defaulting the max requests to 10-12 instead of 6, to cover this use case.

What I meant to say was server, not subdomain. `foo.server.com` and `bar.server.com` are treated separately. I added server throttling back since that comment.

lilleyse commented:
> What's the plan with this w.r.t to HTTP2 and #5316?

As of now these PRs are separate and don't need to interfere with each other.

lilleyse commented:
The code is ready to review again.

My internet connection is not super reliable here so I will hold off on final benchmarks until tomorrow.

lilleyse commented:
Stats on this branch. New takeaways:

* Unthrottled tests are roughly equal to master in terms of speed and number of requests.
* Throttled/prioritized tests perform similar to master and sometimes worse. I'm not too surprised by this since the throttling is still 6-per-server like master but there is an added overhead in cancelling low-priority requests. The best improvements are in the slow-connection "zoom" tests.
* Perceptually, the prioritization helps a lot in loading high detail stuff that is closest to you.
* Testing is not super reliable since there is a bit of a luck factor involved. I often see variance of half a second or more between runs. Sometimes when I re-run a test later I see much better results.
* Why are the # of requests for the unthrottled tests different on this branch than master? I'm not sure, but running master today I see the same numbers.

```
Results on request-scheduler-master

New | HTTP2 server | no-throttle | Static
    5.59 seconds | 513 requests
    6.09 seconds | 489 requests
    5.33 seconds | 478 requests
    5.78 seconds | 504 requests

New | HTTP2 server | throttle 6 per-domain | Static
    4.94 seconds | 319 requests
    4.78 seconds | 330 requests
    4.89 seconds | 332 requests
    4.93 seconds | 323 requests

New | HTTP1 server | no-throttle | Static
    2.89 seconds | 594 requests
    2.74 seconds | 594 requests
    2.73 seconds | 594 requests
    2.61 seconds | 594 requests
    2.79 seconds | 594 requests
    2.56 seconds | 594 requests

New | HTTP1 server | throttle 6 per-domain | Static
    2.74 seconds | 527 requests
    2.89 seconds | 540 requests
    3.03 seconds | 554 requests
    2.91 seconds | 557 requests
    3.04 seconds | 543 requests

New | HTTP1 server | no-throttle | Zoom
    4.14 seconds | 793 requests
    4.13 seconds | 784 requests
    4.18 seconds | 782 requests

New | HTTP1 server | throttle 6 per-domain | Zoom
    3.86 seconds | 712 requests
    3.91 seconds | 754 requests
    3.89 seconds | 701 requests

==========================

New | HTTP1 server | no-throttle | Static | 4G
    24.12 seconds | 624 requests
    24.11 seconds | 624 requests

New | HTTP1 server | throttle 6 per-domain | Static | 4G
    15.04 seconds | 381 requests
    15.37 seconds | 391 requests
    15.34 seconds | 389 requests

New | HTTP1 server | no-throttle | Zoom | 4G
    32.83 seconds | 1026 requests
    35.21 seconds | 1082 requests

New | HTTP1 server | throttle 6 per-domain | Zoom | 4G
    7.40 seconds | 273 requests
    7.48 seconds | 276 requests

New | HTTP2 server | no-throttle | Static | 4G
    16.52 seconds | 446 requests
    17.12 seconds | 460 requests

New | HTTP2 server | throttle 6 per-domain | Static | 4G
    12.57 seconds | 324 requests
    12.08 seconds | 309 requests
```

lilleyse commented:
These are all **4G** tests just to show the contrast better.

**Static view - throttled - master vs. new.**
![oie_oie_overlay 1](https://user-images.githubusercontent.com/915398/26951784-7a63cdb8-4c70-11e7-8aad-58bc14d288a1.gif)

**Static view - not throttled - master vs. new.**
![oie_oie_overlay](https://user-images.githubusercontent.com/915398/26951763-6470bf2a-4c70-11e7-9108-b752a85132cc.gif)

**Zoom view - throttled - master vs. new.**
![oie_oie_overlay 3](https://user-images.githubusercontent.com/915398/26951805-8b807b00-4c70-11e7-8305-970ef20b7fdb.gif)

**Zoom view - not throttled - master vs. new.**
![oie_oie_overlay 2](https://user-images.githubusercontent.com/915398/26951802-86aed34c-4c70-11e7-9f24-1138c29c476b.gif)

lilleyse commented:
The first screenshot shows the benefits the most. My overall impression is that prioritization isn't a huge win for the terrain/imagery system alone since the default traversal is already pretty good. The real benefit will come when we have multiple data sources with different traversal patterns - aka 3D Tiles on the globe.

pjcozzi commented:
@lilleyse this all sounds good to me for now, just leave the `while (true)` if nothing else is cleaner.

We can do more research as we optimize 3D Tiles over time.

@mramato I'd like to merge this soon.  Do you want to look at it again?

lilleyse commented:
> while (true)

I replaced it with a bool, still the same logic though.

mramato commented:
> @mramato I'd like to merge this soon. Do you want to look at it again?

I trust you guys, feel free to merge (unless you are asking because you don't have the time for another once over).

@lilleyse merge in master.
